### ISSUE TITLE
OpenShift unable to deploy because of privileged users

### ISSUE THREAD

**Issue by haarchri** [sentry-kubernetes/charts#67]
hello,

it would be great to ship all statefulsets that it is possible to run in openshift 

for example:
`create Pod sentry-clickhouse-replica-0 in StatefulSet sentry-clickhouse-replica failed error: pods "sentry-clickhouse-replica-0" is forbidden: unable to validate against any security context constraint: [spec.containers[0].securityContext.securityContext.runAsUser: Invalid value: 0: must be in the ranges: [1016890000, 1016899999] spec.containers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed]`

**Comment by DandyDeveloper** [comment]
@haarchri Unfortunately this is actually a dependency issue that would need to be fixed by Clickhouse itself: 

https://github.com/liwenhe1993/charts/blob/master/clickhouse/templates/statefulset-clickhouse.yaml#L94

You can raise an issue there and we may be able to contribute a solution. 

In the meantime, you should be able to configure OpenShift to allow this (If your organisation permits).

**Comment by Mokto** [comment]
Because of this, [this other issue](https://github.com/sentry-kubernetes/charts/pull/50#issuecomment-612444903) and the lack of maintenance, I'm more and more thinking of building a fork of clickhouse inside that repository.

What do you think?

**Comment by DandyDeveloper** [comment]
@Mokto Wouldn't hurt, but it's another thing to manage ;) That's the only downside.

I'm not the most confident with clickhouse, but I'm sure it can't be that hard to follow.

**Comment by haarchri** [comment]
the only open point is clickhouse in the helm chart - i play around with this implementation https://github.com/Uglykoyote/clickhouse will have a look if its possible with this openshift template - if yes i can create a helm chart of it 

the rest is starting up with a values.yaml like this:

```
prefix:

user:
  create: true
  email: test@example.com
  password: aaaa

images:
  sentry:
    repository: getsentry/sentry
    tag: 6e6ed914e24d9121619ccc754325fbba611ad6be
    pullPolicy: IfNotPresent
    # imagePullSecrets: []
  snuba:
    repository: getsentry/snuba
    tag: fdcc3229b24b659a48b4934e4ffc0ccaa6364a37
    pullPolicy: IfNotPresent
    # imagePullSecrets: []

sentry:
  web:
    replicas: 1
    env: {}
    probeInitialDelaySeconds: 10
    resources: {}
    affinity: {}
    nodeSelector: {}
    # tolerations: []
    # podLabels: []

    autoscaling:
      enabled: false
      minReplicas: 2
      maxReplicas: 5
      targetCPUUtilizationPercentage: 50

  worker:
    replicas: 1
    # concurrency: 4
    env: {}
    resources: {}
    affinity: {}
    nodeSelector: {}
    # tolerations: []
    # podLabels: []
  cron:
    env: {}
    resources: {}
    affinity: {}
    nodeSelector: {}
    # tolerations: []
    # podLabels: []
  postProcessForward:
    replicas: 1
    env: {}
    resources: {}
    affinity: {}
    nodeSelector: {}
    # tolerations: []
    # podLabels: []

snuba:
  api:
    replicas: 1
    env: {}
    probeInitialDelaySeconds: 10
    resources: {}
    affinity: {}
    nodeSelector: {}
    # tolerations: []
    # podLabels: []

    autoscaling:
      enabled: false
      minReplicas: 2
      maxReplicas: 5
      targetCPUUtilizationPercentage: 50

  consumer:
    replicas: 1
    env: {}
    resources: {}
    affinity: {}
    nodeSelector: {}
    # tolerations: []
    # podLabels: []
  replacer:
    env: {}
    resources: {}
    affinity: {}
    nodeSelector: {}
    # tolerations: []
    # podLabels: []


hooks:
  enabled: true
  dbInit:
    resources:
      limits:
        memory: 2048Mi
      requests:
        cpu: 300m
        memory: 2048Mi
  snubaInit:
    resources:
      limits:
        cpu: 2000m
        memory: 1Gi
      requests:
        cpu: 700m
        memory: 1Gi

system:
  url: ""
  adminEmail: ""
  secretKey: 'icLq77rCyY_qrMMpXa6TQNjkDV6mU!c'
  public: false #  This should only be used if you’re installing Sentry behind your company’s firewall.

mail:
  backend: dummy # smtp
  useTls: false
  username: ""
  password: ""
  port: 25
  host: ""
  from: ""

symbolicator:
  enabled: false

auth:
  register: true

service:
  name: sentry
  type: ClusterIP
  externalPort: 9000
  annotations: {}
  # externalIPs:
  # - 192.168.0.1
  # loadBalancerSourceRanges: []

github: {} # https://github.com/settings/apps (Create a Github App)
# github:
#   appId: "xxxx"
#   appName: MyAppName
#   clientId: "xxxxx"
#   clientSecret: "xxxxx"
#   privateKey: "-----BEGIN RSA PRIVATE KEY-----\nMIIEpA" !!!! Don't forget a trailing \n
#   webhookSecret:  "xxxxx`"

githubSso: {} # https://github.com/settings/developers (Create a OAuth App)
  # clientId: "xx"
  # clientSecret: "xx"

slack: {}
# slack:
#   clientId:
#   clientSecret:
#   verificationToken:

ingress:
  enabled: false
  # annotations:
  #   kubernetes.io/tls-acme:
  #   certmanager.k8s.io/issuer:
  #   nginx.ingress.kubernetes.io/proxy-body-size:
  #
  # hostname:
  #
  # tls:
  # - secretName:
  #   hosts:

filestore:
  # Set to one of filesystem, gcs or s3 as supported by Sentry.
  backend: filesystem

  filesystem:
    path: /var/lib/sentry/files

    ## Enable persistence using Persistent Volume Claims
    ## ref: http://kubernetes.io/docs/user-guide/persistent-volumes/
    ##
    persistence:
      enabled: true
      ## database data Persistent Volume Storage Class
      ## If defined, storageClassName: <storageClass>
      ## If set to "-", storageClassName: "", which disables dynamic provisioning
      ## If undefined (the default) or set to null, no storageClassName spec is
      ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
      ##   GKE, AWS & OpenStack)
      ##
      # storageClass: "-"
      accessMode: ReadWriteOnce
      size: 10Gi

      ## Whether to mount the persistent volume to the Sentry worker and
      ## cron deployments. This setting needs to be enabled for some advanced
      ## Sentry features, such as private source maps. If you disable this
      ## setting, the Sentry workers will not have access to artifacts you upload
      ## through the web deployment.
      ## Please note that you may need to change your accessMode to ReadWriteMany
      ## if you plan on having the web, worker and cron deployments run on
      ## different nodes.
      persistentWorkers: false

  ## Point this at a pre-configured secret containing a service account. The resulting
  ## secret will be mounted at /var/run/secrets/google
  gcs:
    # credentialsFile: credentials.json
    #  secretName:
    #  bucketName:

  ## Currently unconfigured and changing this has no impact on the template configuration.
  s3: {}
  #  accessKey:
  #  secretKey:
  #  bucketName:
  #  endpointUrl:
  #  signature_version:
  #  region_name:
  #  default_acl:

config:
  configYml: |
    # No YAML Extension Config Given
  sentryConfPy: |
    # No Python Extension Config Given

clickhouse:
  enabled: true

  tabix: 
    enabled: false

  clickhouse:
    image: "yandex/clickhouse-server"
    imageVersion: "20.3"
    imagePullPolicy: "IfNotPresent"

    ## openshift to use pvcs insted of emtyDir()
    persistentVolumeClaim:
      enabled: true

    ## openshift to use pvcs insted of emtyDir()
    dataPersistentVolume: 
      enabled: true
      storage: "100Gi"

kafka:
  zookeeper:
    ## openshift setting
    securityContext:
      enabled: true
      runAsUser: 1016890000
      fsGroup: 1016890000

  ## openshift setting version 9.00
  podSecurityContext:
    runAsUser: 1016890000
    fsGroup: 1016890000

   ## openshift setting version 8.00    
  securityContext:
    enabled: false	
    runAsUser: 1016890000

  enabled: true
  replicaCount: 3
  allowPlaintextListener: true
  defaultReplicationFactor: 3
  offsetsTopicReplicationFactor: 3
  transactionStateLogReplicationFactor: 3
  transactionStateLogMinIsr: 3

  service:
    port: 9092

redis:
  ## openshift setting
  securityContext:
    enabled: false	
    runAsUser: 1016890000
    
  ## Required if the Redis component of this chart is disabled. (Existing Redis)
  hostOverride: ""
  enabled: true
  nameOverride: sentry-redis
  usePassword: false
  # Only used when internal redis is disabled
  # host: redis
  # Just omit the password field if your redis cluster doesn't use password
  # password: redis
  # port: 6379
  master:
    persistence:
      enabled: true

postgresql:
  ## openshift setting
  volumePermissions:
    enabled: false
    securityContext:
      runAsUser: "auto"
      
  ## openshift setting
  securityContext:
    enabled: false

  ## openshift setting  
  shmVolume:
    enabled: false
    chmod:
      enabled: false

  ## dbcs settings
  persistence:
    storageClass: "ebs-gp2-azbeliebig"

  ## Required if the Postgresql component of this chart is disabled. (Existing Postgres)
  hostOverride: ""
  enabled: true
  nameOverride: sentry-postgresql
  postgresqlUsername: postgres
  postgresqlDatabase: sentry
  # Only used when internal PG is disabled
  # postgresqlHost: postgres
  # postgresqlPassword: postgres
  # postgresqlPort: 5432
  # postgresSslMode: require
  replication:
    enabled: false
    slaveReplicas: 2
    synchronousCommit: "on"
    numSynchronousReplicas: 1

rabbitmq:
  ## If disabled, Redis will be used instead as the broker. 
  enabled: false

```

**Comment by Mokto** [comment]
@haarchri I've cloned the clickhouse chart inside our repo.

**Comment by arianf** [comment]
@Mokto The clone still has `privileged: true`, is the clone not ready yet?

https://github.com/sentry-kubernetes/charts/blob/ec588a94520555e512b1b931eb5799e6d08293c9/clickhouse/templates/statefulset-clickhouse.yaml#L109

https://github.com/sentry-kubernetes/charts/blob/ec588a94520555e512b1b931eb5799e6d08293c9/clickhouse/templates/statefulset-clickhouse-replica.yaml#L110

**Comment by Eliesmbr** [comment]
I am very interested in finding a solution for the problem. As @arianf correctly mentioned, privileged: true and runAsUser: 0 is still included in the template of the clickhouse chart.

I guess some Kubernetes instances, including mine, do not allow privilege escalation and runAsUser 0.

**Comment by Eliesmbr** [comment]
@Mokto Is there any update regarding the clone where `privileged: true` is still present ?

**Comment by akloster** [comment]
In order to find out why clickhouse needs to run privileged, I tried to make a sort of underprivileged custom image which I then attempt to run on its own or via the onpremise docker-compose.

This is about as far as I got:

```Dockerfile
FROM yandex/clickhouse-server:21.5.9.4
RUN useradd -rm -d /home/ubuntu -s /bin/bash -g root -u 1000 ubuntu
RUN chgrp -R root /var/lib/clickhouse; chmod g+rwx /var/lib/clickhouse; chmod -R g+rw /var/lib/clickhouse
RUN setcap cap_ipc_lock=+ep /usr/bin/clickhouse
USER ubuntu
```
The setcap doesn't fix anything though.

The container will fail with the following logs:

```
clickhouse_1                                | Processing configuration file '/etc/clickhouse-server/config.xml'.
clickhouse_1                                | Merging configuration file '/etc/clickhouse-server/config.d/docker_related_config.xml'.
clickhouse_1                                | Merging configuration file '/etc/clickhouse-server/config.d/sentry.xml'.
clickhouse_1                                | Logging information to /var/log/clickhouse-server/clickhouse-server.log
clickhouse_1                                | Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
clickhouse_1                                | Logging information to console
clickhouse_1                                | 2021.07.11 08:42:02.325452 [ 1 ] {} <Information> Application: Will watch for the process with pid 33
clickhouse_1                                | 2021.07.11 08:42:02.325639 [ 33 ] {} <Information> Application: Forked a child process to watch
clickhouse_1                                | 2021.07.11 08:42:02.325950 [ 33 ] {} <Information> SentryWriter: Sending crash reports is disabled
clickhouse_1                                | 2021.07.11 08:42:02.363218 [ 33 ] {} <Information> : Starting ClickHouse 21.5.9.4 with revision 54450, build id: 9CBEDA75070EFE2F084EA8DB58AE37149CB6202E, PID 33
clickhouse_1                                | 2021.07.11 08:42:02.363399 [ 33 ] {} <Information> Application: starting up
clickhouse_1                                | 2021.07.11 08:42:02.458367 [ 33 ] {} <Information> Application: Calculated checksum of the binary: DCEA7A4CD0756F7DF804B3305CA633AA, integrity check passed.
clickhouse_1                                | 2021.07.11 08:42:02.458425 [ 33 ] {} <Information> Application: It looks like the process has no CAP_IPC_LOCK capability, binary mlock will be disabled. It could happen due to incorrect ClickHouse package installation. You could resolve the problem manually with 'sudo setcap cap_ipc_lock=+ep /usr/bin/clickhouse'. Note that it will not work on 'nosuid' mounted filesystems.
clickhouse_1                                | 2021.07.11 08:42:02.458526 [ 33 ] {} <Warning> Application: Effective user of the process (ubuntu) does not match the owner of the data (clickhouse).
clickhouse_1                                | 2021.07.11 08:42:02.459048 [ 33 ] {} <Information> SensitiveDataMaskerConfigRead: 1 query masking rules loaded.
clickhouse_1                                | 2021.07.11 08:42:02.463854 [ 33 ] {} <Error> Application: Syntax error: Not a valid floating-point number
clickhouse_1                                | 2021.07.11 08:42:02.463887 [ 33 ] {} <Information> Application: shutting down
clickhouse_1                                | 2021.07.11 08:42:02.464070 [ 34 ] {} <Information> BaseDaemon: Stop SignalListener thread
clickhouse_1                                | 2021.07.11 08:42:02.475845 [ 1 ] {} <Information> Application: Child process exited normally with code 70.
```

**Comment by Iridescens** [comment]
Hi!

Firstly, 
a) Clickhouse is OKD/OpenShift compatible since 21.2 - https://github.com/ClickHouse/ClickHouse/issues/11003.
b) Use Clickhouse from https://hub.docker.com/r/clickhouse/clickhouse-server/ , not yandex/clickhouse - they moved (https://clickhouse.com/#quick-start).
```
clickhouse:
  enabled: true
  clickhouse:
    image: "clickhouse/clickhouse-server"
    imageVersion: "22.3.3.44"
```
c) Using the image above, you **can** and **must** forfeit `initContainer` with `mkdir`. Having `initContainer` makes pod to crashLoopBackoff with init container being unable to `mkdir -p /etc/clickhouse-server/' (this requires root), which is absolutely redundant as later the chart mounts volumes from configMaps onto dedicated directories.
```
1000670000@sentry-clickhouse-0:/$ ls -la /etc/clickhouse-server/
total 72
drwxrwxrwx. 1 clickhouse clickhouse    23 Apr 20 15:18 .
drwxr-xr-x. 1 root       root          43 Apr 20 15:18 ..
drwxrwsrwx. 3 root       1000670000    78 Apr 20 15:18 config.d
-rw-rw-rw-. 1 clickhouse clickhouse 61569 Apr  5 14:45 config.xml
drwxrwsrwx. 3 root       1000670000    79 Apr 20 15:18 metrica.d
drwxrwsrwx. 3 root       1000670000    76 Apr 20 15:18 users.d
-rw-rw-rw-. 1 clickhouse clickhouse  6248 Apr  5 14:45 users.xml
```
Secondly.
running this modified chart, deployment `sentry-snuba-api` produces
`error: failed switching to "snuba": operation not permitted` which is a blocking issue (https://github.com/getsentry/self-hosted/issues/365) since `gosu` is not permitted in OKD/OpenShift without extended permissions.

Thirdly,
unfortunately modified chart still requires `nonroot` scc applied to `default`, `sentry-sentry-redis`, `sentry-rabbitmq`, `sentry-kafka` service accounts. Default SA can also be stripped of `nonroot` when/if all elements could use dedicated SA. This is caused by underlying charts running with users 1001, which is outside of OKD/OpenShift allowed UID range and requires creation of dedicated scc.

**Comment by Iridescens** [comment]
Well, after testing `sentry-snuba-api` with `securityContext.runAsUser:0` and `scc` `privileged`, usage of CH 22.3.3.44 with stock sentry migrations is a no-go as snuba-migrate job is failing horribly due to non-existent table in CH.
```
2022-04-25 23:08:32,731 Attempting to connect to Clickhouse cluster sentry-clickhouse:9000 (attempt 0)
2022-04-25 23:08:32,740 Connecting. Database: default. User: default
2022-04-25 23:08:32,741 Connecting to sentry-clickhouse:9000
2022-04-25 23:08:32,748 Connected to ClickHouse server version 22.3.3, revision: 54455
2022-04-25 23:08:32,749 Query: SELECT version()
2022-04-25 23:08:32,750 Block "" send time: 0.000218
2022-04-25 23:08:32,756 Query: SELECT group, migration_id, status FROM migrations_dist FINAL WHERE group IN ('system', 'events', 'transactions', 'discover', 'outcomes', 'metrics', 'sessions')
2022-04-25 23:08:32,757 Block "" send time: 0.000130
2022-04-25 23:08:32,764 Running migration: 0001_migrations
2022-04-25 23:08:32,765 Connecting. Database: default. User: default
2022-04-25 23:08:32,765 Connecting to sentry-clickhouse:9000
2022-04-25 23:08:32,772 Connected to ClickHouse server version 22.3.3, revision: 54455
2022-04-25 23:08:32,772 Query: select host_name, port, shard_num, replica_num from system.clusters where cluster='sentry-clickhouse'
2022-04-25 23:08:32,773 Block "" send time: 0.000157
2022-04-25 23:08:32,780 Query: select host_name, port, shard_num, replica_num from system.clusters where cluster='sentry-clickhouse'
2022-04-25 23:08:32,780 Block "" send time: 0.000137
2022-04-25 23:08:32,784 Finished: 0001_migrations
2022-04-25 23:08:32,785 Connecting. Database: default. User: default
2022-04-25 23:08:32,785 Connecting to sentry-clickhouse:9000
2022-04-25 23:08:32,792 Connected to ClickHouse server version 22.3.3, revision: 54455
2022-04-25 23:08:32,793 Query: SELECT version FROM migrations_dist FINAL WHERE group = 'system' AND migration_id = '0001_migrations';
2022-04-25 23:08:32,793 Block "" send time: 0.000123
Traceback (most recent call last):
  File "/usr/src/snuba/snuba/clickhouse/native.py", line 187, in execute
    result_data = query_execute()
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 286, in execute
    rv = self.process_ordinary_query(
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 479, in process_ordinary_query
    return self.receive_result(with_column_types=with_column_types,
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 136, in receive_result
    return result.get_result()
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/result.py", line 50, in get_result
    for packet in self.packet_generator:
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 152, in packet_generator
    packet = self.receive_packet()
  File "/usr/local/lib/python3.8/site-packages/clickhouse_driver/client.py", line 169, in receive_packet
    raise packet.exception
clickhouse_driver.errors.ServerException: Code: 60.
DB::Exception: Table default.migrations_dist doesn't exist.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/snuba", line 33, in <module>
    sys.exit(load_entry_point('snuba', 'console_scripts', 'snuba')())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/src/snuba/snuba/cli/migrations.py", line 64, in migrate
    runner.run_all(force=force)
  File "/usr/src/snuba/snuba/migrations/runner.py", line 157, in run_all
    self._run_migration_impl(migration_key, force=force)
  File "/usr/src/snuba/snuba/migrations/runner.py", line 215, in _run_migration_impl
    migration.forwards(context, dry_run)
  File "/usr/src/snuba/snuba/migrations/migration.py", line 149, in forwards
    update_status(Status.COMPLETED)
  File "/usr/src/snuba/snuba/migrations/runner.py", line 313, in _update_migration_status
    next_version = self._get_next_version(migration_key)
  File "/usr/src/snuba/snuba/migrations/runner.py", line 328, in _get_next_version
    result = self.__connection.execute(
  File "/usr/src/snuba/snuba/clickhouse/native.py", line 243, in execute
    raise ClickhouseError(e.message, code=e.code) from e
snuba.clickhouse.errors.ClickhouseError: DB::Exception: Table default.migrations_dist doesn't exist.
```

**Comment by pauvos** [comment]
@Iridescens , thanks for pointing out the change in clickhouse version 21.2. Recent versions indeed start but for some reason the migrations fail to run.

I didn't have any luck with the clickhouse/clickhouse-server images but hacking the entryscript of yandex/clickhouse-server with kustomize works:

```yaml
command:
  - /bin/bash
  - -c
  - |
    export SHARD=${HOSTNAME##*-}
    cp /entrypoint.sh /tmp/entrypoint.sh
    sed -i '/elif [ "$(stat -c %u "$dir")" != "$USER" ]/,+2d' /tmp/entrypoint.sh
    /tmp/entrypoint.sh
```

This removes the problematic permission check from the 20.8 entryscript: https://github.com/ClickHouse/ClickHouse/blob/20.8/docker/server/entrypoint.sh#L64

So after...
* changing the snuba api entrypoint from `gosu snuba snuba api` to `snuba api` (#644)
* changing the clickhouse entrypoint (see above)
* removing the clickhouse initContainer (#609)
* adding securityContext to all hook Jobs (#643)

... everything appears to be working without any errors in the logs.

**Comment by cortopy** [comment]
It's been an ardous task, but I managed to install sentry chart without root.

For anyone running secure clusters, this is what I needed to do:
- Use clickhouse version `21.8.15.7` using [clichouse operator](https://github.com/Altinity/clickhouse-operator). This is important as newer versions of clickhouse don't require root. Beware though, as per this [issue](https://github.com/sentry-kubernetes/charts/issues/677), versions newer than this one don't work. EDIT: actually it has to be `20.8.9.6` from `yandex/clickhouse-server` as per https://github.com/getsentry/snuba/issues/1837#issuecomment-829148357
- Add `CLICHOUSE_PORT` environment variable with value of "9000" to all snuba containers. This is because this variable is wrongly set to the full url when using external clickhouse and the python `int` transpose function will fatally fail 
- Add `securityContext` to all pods, as they all run with a named user by default and kubernetes doesn't have a way to know if it's running as non-root:
```yaml
securityContext:
   runAsUser: 1000
```
- Pipe the output of the helm chart to kustomise to add the `securityContext` to the three cron jobs as the helm templates have none
